### ISSUE TITLE
Subjects still not retiring correctly

### ISSUE THREAD

**Issue by mkosmala** [zooniverse/Panoptes#1748]
I thought this was fixed long ago. But I'm now looking at the data generated from last Tuesday, when @mrniaboc sent out the newsletter about the Season Spotter spring challenge. All subjects are supposed to retire with 5 classifications. But they don't necessarily. At first I thought maybe this is because of the way subjects are served in sets -- if classification 5 comes in while someone else is working on the same subject, we might occasionally end up with 6 classifications. But that's not the (only) issue. I also have subjects like subject 589803 for which I have 7 classifications, with created_at stamps of:
2015-12-02 20:17:40 UTC
2015-12-04 15:37:26 UTC
2015-12-22 20:02:42 UTC
2015-12-22 20:48:34 UTC
2015-12-23 15:00:59 UTC
2016-02-02 20:09:06 UTC
2016-03-23 01:11:49 UTC
This subject should have retired in December. Why was it sent out again in February? And again last week?

Please look into this. It's costing us a TON of volunteer effort. (I estimate that of the 30,000 classifications that came in last Tuesday, ~6,000  (i.e. 20%) of them were for subjects that should have been retired.)

**Comment by mschwamb** [comment]
@mkosmala Are you sure those users classifying in February didn't classify the image when it was in a retired state after they've seen everything else? The interface will someone new images that are retired but they haven't seen with the retired flag on it, This might be more related to https://github.com/zooniverse/Panoptes-Front-End/issues/1050

**Comment by mkosmala** [comment]
@mschwamb I thought of this, but it is definitely not what is going on. Our most prolific classifier hasn't even seen 25% of the subjects in our most-complete workflow and a double-check of her classifications shows that 7 of her most recent 10 classifications in that workflow were on subjects that hadn't reached retirement yet. (One of those other 3, sadly, has been classified by 10 people now! And we still have ~10% of our subjects that have fewer than 5 classifications...)

**Comment by camallen** [comment]
Hey @mkosmala 

the subject retired just fine on 23rd Dec 2015
```
workflow.id
=> 174
workflow.finished_at
=> Wed, 23 Mar 2016 08:45:52 UTC +00:00
subject_workflow.retired_at     
=> Wed, 23 Dec 2015 15:00:59 UTC +00:00
```

Email on 3rd March 2016, you asked me to ensure only workflow_id: #174 was active, the rest were set to inactive in preparation for your promotion?

The classification on the 2016-02-02 20:09:06 UTC is tricky to debug but luckily the user subject queue is still in a state from their classifying session on 2nd Feb 2016. All the data in it is retired for some reason, most likely the workflow was close to completion and the selector couldn't find any unseen, non-retired data for that user ([code here](https://github.com/zooniverse/Panoptes/blob/6237b1df643c8464ecb1cab868088a2bb3ada083/lib/subjects/set_member_subject_selector.rb#L31)) so it was showing the user retired data that they haven't seen (this is by design instead of showing them nothing).

The classification on the 2016-03-23 01:11:49 UTC was the same and very close to the retirement date time so most of the data was complete. Having more data to select from would have alleviated this. E.g. you had 1 active subject set that basically ran out of non-retired data when you had a lot of users, if you had more active data it would have selected from this pool instead of trying to select from mostly retired data.

Happy to hash out the details more on this if needed.

You did ask us to activate data (email) on the 23rd at 00:55am and again on the 24th. I think this mostly a data management issue under active promotion and linked to #1707. Instead of just classifying the active linked pool of data we really need to bring new pools online when we reach a threshold limit. Thoughts on that issue are appreciated.

Please close the issue if you're happy enough with the answer (if not the result).

**Comment by mkosmala** [comment]
Hey @camallen, thanks for looking into this. On 2nd Feb, there were no workflows (or subject sets) that were close to being fully retired. The one most complete was between 1/3 and 1/2 done, I'm pretty sure. (I can reconstruct it exactly, if you need.) 

And it's baffling to me why this has happened so frequently and across all 7 workflows (most of which are not close to being fully retired). I really don't think the problem is that it's only people who have "run out" of non-retired subjects who are getting retired subjects to classify. I think something else is going on. 

I can go and pull out some more examples and we can try to reconstruct what happened, and perhaps find a common thing going on... thoughts?

**Comment by camallen** [comment]
Please do, the more data we have to figure out what was up the better as it is hard trying to reconstruct the state of the system in the past to debug.

**Comment by camallen** [comment]
@mkosmala see the fix in #1767 for a probably cause of the issues you are seeing. 

The code was most likely signalling the workflow finished state based on incorrect counts against changing subject data counts (linking / unlinking data). Looking to get that fixed asap.
